src/components/modebar/index.js

```diff
@@ -97,18 +122,31 @@ proto.createButton = function (config) {
     button.setAttribute('rel', 'tooltip');
     button.className = 'modebar-btn';
 
-    if (config.attr !== undefined) button.setAttribute('data-attr', config.attr);
-    if (config.val !== undefined) button.setAttribute('data-val', config.val);
-    button.setAttribute('data-title', config.title);
+    button.setAttribute('data-title', config.title || '');
     button.setAttribute('data-gravity', config.gravity || 'n');
-    button.addEventListener('click', function () {
-            config.click.apply(_this, arguments);
+
+    if(config.attr !== undefined) button.setAttribute('data-attr', config.attr);
+
+    var val = config.val;
+    if(val !== undefined) {
+        if(typeof val === 'function') val = val(this.graphInfo);
+        button.setAttribute('data-val', val);
+    }
+
+    var click = config.click;
+    if(typeof click !== 'function') {
+        throw new Error('must provide button \'click\' function in button config');
+    }
+    else {
+        button.addEventListener('click', function(ev) {
+            config.click(_this, ev);
```

Comment:
(but the way to do it would be to bind the click handler to modebar inside the modebar machinery)
```
button.addEventListener('click', configCallback.bind(this))
```
then in the API we can let folks know when writing their handler they have access to modebar functionality as 
```
this.doSomethingWithTheOtherButtons()
```